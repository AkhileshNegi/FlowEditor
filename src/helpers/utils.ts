import { ComponentClass, SFC, ReactElement } from 'react';
import { ShallowWrapper, ReactWrapper } from 'enzyme';
import { FlowDefinition } from '../flowTypes';

const request = require('sync-request');

const XRegExp = require('xregexp');

const SNAKED_CHARS = XRegExp('[^\\p{Letter}\\d]+');

interface BoolMap {
    [key: string]: boolean;
}

/**
 * Turns a string array into a bool map for constant lookup
 * @param {string[]} array - an array of strings, e.g. contact fields
 * @returns {object} A map of each string
 */
export function toBoolMap(array: string[]): BoolMap {
    return array.reduce((map: BoolMap, item: string) => {
        map[item] = true;
        return map;
    }, {});
}

/**
 * Inserts commas into numbers where appropriate for better readability
 * @param {number} value - A number, e.g. 10000
 * @returns {string} A comma-separated string, e.g. 10,000
 */
export function addCommas(value: number): string {
    return value.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ',');
}

/**
 * Replaces spaces with underscores
 * @param {string} value - A space-separated string to be snaked, e.g. a raw flow field name ('my flow field')
 * @returns {string} A snaked string, e.g. 'my_flow_field'
 */
export function snakify(value: string): string {
    return XRegExp.replace(value.toLowerCase().trim(), SNAKED_CHARS, '_', 'all');
}

/**
 * Returns a React component's name so we can attach it to a HOC's displayName property and view it in RDT
 * @param {ComponentClass | SFC} Component - A React component
 * @returns {string} The component's name
 */
export function getDisplayName(HOCName: string, Component: ComponentClass | SFC): string {
    const ComponentDisplayName = Component.displayName || Component.name || 'Component';
    return `${HOCName}(${ComponentDisplayName})`;
}

function getFlow(name: string): FlowDefinition {
    const definition = request('GET', 'base/test_flows/' + name + '.json').getBody();
    return JSON.parse(definition) as FlowDefinition;
}

export function getFavorites(): FlowDefinition {
    return getFlow('favorites');
}

export function getTest(): FlowDefinition {
    return getFlow('a4f64f1b-85bc-477e-b706-de313a022979');
}

export function dump(object: any) {
    console.log(JSON.stringify(object, null, 1));
}

export const DATA_SPEC_ATTRIBUTE_NAME: string = 'data-spec';

/**
 * NOTE: borrowed from EventBrite: https://github.com/eventbrite/javascript/blob/master/react/testing.md#finding-nodes
 * Finds all instances of components in the rendered `componentWrapper` that are DOM components
 * with the `data-spec` attribute matching `name`.
 * @param {ReactWrapper|ShallowWrapper} componentWrapper - Rendered componentWrapper (result of mount, shallow, or render)
 * @param  {string} snacName - Name of `data-spec` attribute value to find
 * @param {string|Function} (Optional) typeFilter - (Optional) Expected type of the wrappers (defaults to all HTML tags)
 * @returns {ReactWrapper|ReactWrapper[]|ShallowWrapper|ShallowWrapper[]} matching DOM components
 */
export function getSpecWrapper(
    componentWrapper: ReactWrapper | ShallowWrapper,
    specName: string
): ShallowWrapper | ReactWrapper {
    return componentWrapper.find(`[${DATA_SPEC_ATTRIBUTE_NAME}="${specName}"]`);
}

/**
 * Returns true if a given UUID matches v4 format
 * @param {string} uuid - A version 4 UUID (no braces, uppercase OK)
 * @returns {boolean}
 */
export function validUUID(uuid: string): boolean {
    const regex = /^[0-9A-F]{8}-[0-9A-F]{4}-4[0-9A-F]{3}-[89AB][0-9A-F]{3}-[0-9A-F]{12}$/i;
    return regex.test(uuid);
}
