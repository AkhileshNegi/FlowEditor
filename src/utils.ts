const XRegExp = require('xregexp');

const SNAKED_CHARS = XRegExp('[^\\p{Letter}\\d]+');

interface BoolMap {
    [key: string]: boolean;
}

/**
 * Turns a string array into a bool map for constant lookup
 * @param {string[]} array - an array of strings, e.g. contact fields
 * @returns {BoolMap} A map of each string
 */
export function toBoolMap(array: string[]): BoolMap {
    return array.reduce((map: BoolMap, item: string) => {
        map[item] = true;
        return map;
    }, {});
}

export function addCommas(value: number): string {
    return value.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ',');
}

/**
 * Replaces spaces with underscores
 * @param {string} value - A space-separated string to be snaked, e.g. a raw flow field name ('my flow field')
 * @returns {string} A snaked string, e.g. 'my_flow_field'
 */
export function snakify(value: string): string {
    return XRegExp.replace(value.toLowerCase(), SNAKED_CHARS, '_', 'all');
}
