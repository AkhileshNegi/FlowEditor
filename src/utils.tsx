/** Turns a string array into a bool map */
export function toBoolMap(array: string[]): {[item: string]: boolean} {
    var map: {[item: string]: boolean} = {};
    for (let item of array) {
        map[item] = true;
    }
    return map;
}