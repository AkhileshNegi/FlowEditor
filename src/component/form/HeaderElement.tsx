import { react as bindCallbacks } from 'auto-bind';
import * as React from 'react';

import { hasErrorType, renderIf } from '../../utils';
import ConnectedTextInputElement, { HTMLTextElement } from '../form/TextInputElement';
import * as styles from '../routers/Webhook.scss';
import FormElement from './FormElement';

export interface Header {
    uuid: string;
    name: string;
    value: string;
}

export interface HeaderElementProps {
    name: string;
    header: Header;
    index: number;
    onRemove: (header: HeaderElement) => void;
    onChange: (header: HeaderElement) => void;
    empty?: boolean;
}

interface HeaderElementState {
    name: string;
    value: string;
    errors: string[];
}

export const headerContainerSpecId = 'header-container';
export const nameContainerSpecId = 'name-container';
export const valueConatainerSpecId = 'value-container';
export const removeIcoSpecId = 'remove-icon';

export const HEADER_NAME_ERROR = 'HTTP headers must have a name';
export const NAME_PLACEHOLDER = 'Header Name';
export const VALUE_PLACEHOLDER = 'Value';

export default class HeaderElement extends React.Component<HeaderElementProps, HeaderElementState> {
    constructor(props: HeaderElementProps) {
        super(props);

        const name = this.props.header.name || '';
        const value = this.props.header.value || '';

        this.state = {
            name,
            value,
            errors: []
        };

        bindCallbacks(this, {
            include: [/^on/]
        });
    }

    private onChangeName({
        currentTarget: { value }
    }: React.SyntheticEvent<HTMLTextElement>): void {
        this.setState(
            {
                name: value
            },
            () => this.props.onChange(this)
        );
    }

    private onChangeValue({
        currentTarget: { value }
    }: React.SyntheticEvent<HTMLTextElement>): void {
        this.setState(
            {
                value
            },
            () => this.props.onChange(this)
        );
    }

    private onRemove(): void {
        this.props.onRemove(this);
    }

    public validate(): boolean {
        const errors = [];

        const needsHeaderName =
            this.state.value.trim().length > 0 && this.state.name.trim().length === 0;

        if (needsHeaderName) {
            errors.push(HEADER_NAME_ERROR);
        }

        this.setState({ errors });

        return errors.length === 0;
    }

    private getRemoveIco(): JSX.Element {
        return renderIf(this.props.index !== 0 && !this.props.empty)(
            <div className={styles.removeIco} onClick={this.onRemove} data-spec={removeIcoSpecId}>
                <span className="fe-remove" />
            </div>
        );
    }

    public render(): JSX.Element {
        const hasHeaderError = hasErrorType(this.state.errors, [/headers/]);
        const removeIco: JSX.Element = this.getRemoveIco();
        return (
            <FormElement name={this.props.name} errors={this.state.errors}>
                <div className={styles.header} data-spec={headerContainerSpecId}>
                    <div className={styles.header_name} data-spec={nameContainerSpecId}>
                        <ConnectedTextInputElement
                            placeholder={NAME_PLACEHOLDER}
                            name="name"
                            onChange={this.onChangeName}
                            value={this.state.name}
                            showInvalid={hasHeaderError}
                        />
                    </div>
                    <div className={styles.header_value} data-spec={valueConatainerSpecId}>
                        <ConnectedTextInputElement
                            placeholder={VALUE_PLACEHOLDER}
                            name="value"
                            onChange={this.onChangeValue}
                            value={this.state.value}
                            autocomplete={true}
                        />
                    </div>
                    {removeIco}
                </div>
            </FormElement>
        );
    }
}
