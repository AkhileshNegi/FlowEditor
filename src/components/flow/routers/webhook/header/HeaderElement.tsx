import { react as bindCallbacks } from 'auto-bind';
import { getAllErrors } from 'components/flow/actions/helpers';
import { HeaderEntry } from 'components/flow/routers/webhook/WebhookRouterForm';
import styles from 'components/flow/routers/webhook/WebhookRouterForm.module.scss';
import FormElement from 'components/form/FormElement';
import TextInputElement from 'components/form/textinput/TextInputElement';
import * as React from 'react';
import { StringEntry, ValidationFailure } from 'store/nodeEditor';
import { HeaderName, validate } from 'store/validators';

// TODO: move this into webhook router component
export interface Header {
  uuid: string;
  name: string;
  value: string;
}

export interface HeaderElementProps {
  entry: HeaderEntry;
  index: number;
  onRemove: (header: Header) => void;
  onChange: (header: Header, validationFailures: ValidationFailure[]) => void;
  empty?: boolean;
}

interface HeaderElementState {
  name: StringEntry;
  value: StringEntry;
}

export const headerContainerSpecId = 'header-container';
export const nameContainerSpecId = 'name-container';
export const valueConatainerSpecId = 'value-container';
export const removeIcoSpecId = 'remove-icon';

export const HEADER_NAME_ERROR = 'HTTP headers must have a name';
export const NAME_PLACEHOLDER = 'Header Name';
export const VALUE_PLACEHOLDER = 'Value';

export default class HeaderElement extends React.Component<HeaderElementProps, HeaderElementState> {
  constructor(props: HeaderElementProps) {
    super(props);

    const header = this.props.entry.value;
    const name = header.name || '';
    const value = header.value || '';

    this.state = {
      name: { value: name },
      value: { value }
    };

    bindCallbacks(this, {
      include: [/^on/, /^handle/]
    });
  }

  private getHeader(): Header {
    return {
      name: this.state.name.value,
      value: this.state.value.value,
      uuid: this.props.entry.value.uuid
    };
  }

  private handleChangeName(value: string): void {
    const name = validate('Header name', value, [HeaderName]);
    this.setState({ name: { value: name.value } }, () =>
      this.props.onChange(
        this.getHeader(),
        getAllErrors(this.state.value).concat(getAllErrors(name))
      )
    );
  }

  private handleChangeValue(value: string): void {
    this.setState({ value: { value } }, () => {
      const name = validate('Header name', this.state.name.value, [HeaderName]);
      this.props.onChange(
        this.getHeader(),
        getAllErrors(this.state.value).concat(getAllErrors(name))
      );
    });
  }

  private handleRemove(): void {
    this.props.onRemove(this.getHeader());
  }

  private getRemoveIco(): JSX.Element {
    return (
      <div className={styles.remove_ico} onClick={this.handleRemove} data-spec={removeIcoSpecId}>
        <span className="fe-x" />
      </div>
    );
  }

  public render(): JSX.Element {
    const removeIco: JSX.Element = this.getRemoveIco();
    return (
      <FormElement name="Header" entry={this.props.entry}>
        <div className={styles.header} data-spec={headerContainerSpecId}>
          <div className={styles.header_name} data-spec={nameContainerSpecId}>
            <TextInputElement
              placeholder={NAME_PLACEHOLDER}
              name="name"
              onChange={this.handleChangeName}
              entry={this.state.name}
            />
          </div>
          <div className={styles.header_value} data-spec={valueConatainerSpecId}>
            <TextInputElement
              placeholder={VALUE_PLACEHOLDER}
              name="value"
              onChange={this.handleChangeValue}
              entry={this.state.value}
              onFieldFailures={(validationFailures: ValidationFailure[]) => {
                const name = validate('Header name', this.state.name.value, [HeaderName]);
                this.props.onChange(
                  this.getHeader(),
                  validationFailures.concat(getAllErrors(name))
                );
              }}
              autocomplete={true}
            />
          </div>
          {removeIco}
        </div>
      </FormElement>
    );
  }
}
