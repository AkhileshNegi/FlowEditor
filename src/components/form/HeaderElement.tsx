import { react as bindCallbacks } from 'auto-bind';
import * as React from 'react';
import * as styles from '~/components/flow/routers/Webhook.scss';
import ConnectedTextInputElement from '~/components/form/textinputelement/TextInputElement';
import { renderIf } from '~/utils';

import FormElement from './FormElement';

export interface Header {
    uuid: string;
    name: string;
    value: string;
}

export interface HeaderElementProps {
    name: string;
    header: Header;
    index: number;
    onRemove: (header: HeaderElement) => void;
    onChange: (header: HeaderElement) => void;
    empty?: boolean;
}

interface HeaderElementState {
    name: string;
    value: string;
}

export const headerContainerSpecId = 'header-container';
export const nameContainerSpecId = 'name-container';
export const valueConatainerSpecId = 'value-container';
export const removeIcoSpecId = 'remove-icon';

export const HEADER_NAME_ERROR = 'HTTP headers must have a name';
export const NAME_PLACEHOLDER = 'Header Name';
export const VALUE_PLACEHOLDER = 'Value';

export default class HeaderElement extends React.Component<HeaderElementProps, HeaderElementState> {
    constructor(props: HeaderElementProps) {
        super(props);

        const name = this.props.header.name || '';
        const value = this.props.header.value || '';

        this.state = {
            name,
            value
        };

        bindCallbacks(this, {
            include: [/^on/, /^handle/]
        });
    }

    private handleChangeName(name: string): void {
        this.setState({ name }, () => this.props.onChange(this));
    }

    private handleChangeValue(value: string): void {
        this.setState({ value }, () => this.props.onChange(this));
    }

    private onRemove(): void {
        this.props.onRemove(this);
    }

    /* 
    // TOOO: this one needs cross field validation
    public validate(): boolean {
        const errors = [];

        const needsHeaderName =
            this.state.value.trim().length > 0 && this.state.name.trim().length === 0;

        if (needsHeaderName) {
            errors.push(HEADER_NAME_ERROR);
        }

        this.setState({ errors });

        return errors.length === 0;
    }
    */

    private getRemoveIco(): JSX.Element {
        return renderIf(this.props.index !== 0 && !this.props.empty)(
            <div className={styles.removeIco} onClick={this.onRemove} data-spec={removeIcoSpecId}>
                <span className="fe-remove" />
            </div>
        );
    }

    public render(): JSX.Element {
        const hasHeaderError = false; // hasErrorType(this.state.errors, [/headers/]);
        const removeIco: JSX.Element = this.getRemoveIco();
        return (
            <FormElement name={this.props.name}>
                <div className={styles.header} data-spec={headerContainerSpecId}>
                    <div className={styles.header_name} data-spec={nameContainerSpecId}>
                        <ConnectedTextInputElement
                            placeholder={NAME_PLACEHOLDER}
                            name="name"
                            onChange={this.handleChangeName}
                            entry={{ value: this.state.name }}
                            showInvalid={hasHeaderError}
                        />
                    </div>
                    <div className={styles.header_value} data-spec={valueConatainerSpecId}>
                        <ConnectedTextInputElement
                            placeholder={VALUE_PLACEHOLDER}
                            name="value"
                            onChange={this.handleChangeValue}
                            entry={{ value: this.state.value }}
                            autocomplete={true}
                        />
                    </div>
                    {removeIco}
                </div>
            </FormElement>
        );
    }
}
