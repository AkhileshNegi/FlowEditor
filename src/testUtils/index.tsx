// TODO: Remove use of Function
// tslint:disable:ban-types
import { mount, ReactWrapper, shallow, ShallowWrapper } from 'enzyme';
import { object } from 'prop-types';
import * as React from 'react';
import * as config from '../../__test__/config';
import { ConfigProviderContext, endpointsPT, flowPT, languagesPT } from '../config';
import { FlowDefinition, FlowEditorConfig } from '../flowTypes';
import { AppState, createStore, initialState } from '../store';
import { getBaseLanguage } from '../utils';

export interface Resp {
    results: Array<{ [key: string]: any }>;
    total?: number;
}

export interface QueryString {
    [key: string]: string;
}

const initialContext: ConfigProviderContext = {
    endpoints: (config as FlowEditorConfig).endpoints,
    languages: (config as FlowEditorConfig).languages,
    flow: (config as FlowEditorConfig).flow
};

const initialTestState: AppState = {
    ...initialState,
    ...{
        flowContext: {
            ...initialState.flowContext,
            definition: require('../../__test__/flows/colors.json') as FlowDefinition
        }
    },
    ...{
        flowEditor: {
            ...initialState.flowEditor,
            editorUI: {
                ...initialState.flowEditor.editorUI,
                language: getBaseLanguage((config as FlowEditorConfig).languages)
            }
        }
    }
};

/**
 * Compose setup method for component tests
 */
export const createSetup = <P extends {}>(
    Component: React.ComponentClass | React.SFC,
    baseProps: P = {} as any,
    baseDuxState: AppState | Partial<AppState> = initialState,
    baseContext: ConfigProviderContext = initialContext,
    connectedChildren: boolean = false
) => (
    shallowRender: boolean = true,
    propOverrides: P | Partial<P> = {},
    duxStateOverrides = {} as AppState,
    contextOverrides = {} as ConfigProviderContext | Partial<ConfigProviderContext>,
    childContextTypeOverrides: { [key: string]: Function } = {}
) => {
    // Waiting on https://github.com/Microsoft/TypeScript/pull/1328
    const props = Object.assign({}, baseProps, propOverrides);
    const context = connectedChildren
        ? Object.assign({}, baseContext, {
              store: createStore(Object.assign({}, baseDuxState, duxStateOverrides))
          })
        : baseContext;
    // tslint:disable-next-line:ban-types
    const wrapper = (shallowRender ? (shallow as Function) : (mount as Function))(
        <Component {...props} />,
        {
            context,
            childContextTypes: Object.assign(
                {},
                {
                    store: object,
                    endpoints: endpointsPT,
                    languages: languagesPT,
                    flow: flowPT
                },
                childContextTypeOverrides
            )
        }
    );

    return {
        wrapper,
        props,
        context,
        instance: wrapper.instance()
    };
};
export const createSpy = (obj: Object | React.ComponentClass) => (instanceMethod: string) =>
    jest.spyOn((obj as React.ComponentClass).prototype || obj, instanceMethod as any);

/**
 * Wait for promises in queue to resolve
 */
export const flushPromises = () => new Promise(resolve => setImmediate(resolve));

/**
 * Restore spy mocks (distinct from mocks created w/ jest.fn()).
 * Use to declaratively restore multiple spy mocks,
 * otherwise just call .mockRestore() on lone spy.
 */
export const restoreSpies = (...spies: jest.SpyInstance[]) =>
    spies.forEach(spy => spy.mockRestore());

/**
 * NOTE: borrowed from EventBrite: https://github.com/eventbrite/javascript/blob/master/react/testing.md#finding-nodes
 * Finds all instances of components in the rendered `componentWrapper` that are DOM components
 * with the `data-spec` attribute matching `name`.
 * @param {ReactWrapper|ShallowWrapper} componentWrapper - Rendered componentWrapper (result of mount, shallow, or render)
 * @param  {string} snacName - Name of `data-spec` attribute value to find
 * @param {string|Function} (Optional) typeFilter - (Optional) Expected type of the wrappers (defaults to all HTML tags)
 * @returns {ReactWrapper|ReactWrapper[]|ShallowWrapper|ShallowWrapper[]} Matching DOM components
 */
export const getSpecWrapper = (
    componentWrapper: ReactWrapper<{}, {}> | ShallowWrapper<{}, {}>,
    specName: string,
    attributeName: string = 'data-spec'
): any => {
    return componentWrapper.find(`[${attributeName}="${specName}"]`);
};

const DEBUG = false;

export const timeStart = (name: string) => {
    if (DEBUG) {
        console.time(name);
    }
};

export const timeEnd = (name: string) => {
    if (DEBUG) {
        console.timeEnd(name);
    }
};
