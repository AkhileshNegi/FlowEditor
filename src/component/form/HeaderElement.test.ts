import { composeComponentTestUtils, getSpecWrapper, setMock } from '../../testUtils';
import { set, setFalse } from '../../utils';
import HeaderElement, {
    Header,
    headerContainerSpecId,
    HeaderElementProps,
    NAME_PLACEHOLDER,
    nameContainerSpecId,
    removeIcoSpecId,
    VALUE_PLACEHOLDER,
    valueConatainerSpecId
} from './HeaderElement';

const headers: Header[] = [
    {
        uuid: '00c4498e-1c9e-4c26-aa7c-f81d13573129',
        name: 'Content-Type',
        value: 'application/json'
    },
    {
        uuid: '3e7d8366-22ff-452b-953e-3a9a88b5d72c ',
        name: 'Authorization',
        value:
            'Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiYWRtaW4iOnRydWV9.TJVA95OrM7E2cBab30RMHrHDcEfxjoYZgeFONFh7HgQ'
    },
    { uuid: '236f2392-e576-4254-8c28-db175510c6a8', name: '', value: '' }
];

const baseProps: HeaderElementProps = {
    name: `header_${headers[0].uuid}`,
    header: headers[0],
    index: 0,
    onRemove: jest.fn(),
    onChange: jest.fn()
};

const { setup, spyOn } = composeComponentTestUtils(HeaderElement, baseProps);

describe(HeaderElement.name, () => {
    describe('render', () => {
        it('should render self, children with base props', () => {
            const { wrapper, props, instance } = setup();
            const inputs = wrapper.find('Connect(TextInputElement)');

            expect(wrapper.find('FormElement').props()).toEqual(
                expect.objectContaining({
                    name: props.name
                })
            );

            expect(getSpecWrapper(wrapper, headerContainerSpecId).hasClass('header')).toBeTruthy();

            expect(
                getSpecWrapper(wrapper, nameContainerSpecId).hasClass('header_name')
            ).toBeTruthy();

            expect(inputs.at(0).props()).toEqual({
                placeholder: NAME_PLACEHOLDER,
                name: 'name',
                onChange: instance.handleChangeName,
                entry: { value: props.header.name },
                showInvalid: false
            });

            expect(
                getSpecWrapper(wrapper, valueConatainerSpecId).hasClass('header_value')
            ).toBeTruthy();

            expect(inputs.at(1).props()).toEqual({
                placeholder: VALUE_PLACEHOLDER,
                name: 'value',
                onChange: instance.handleChangeValue,
                entry: { value: props.header.value },
                autocomplete: true
            });

            expect(getSpecWrapper(wrapper, removeIcoSpecId).exists()).toBeFalsy();

            expect(wrapper).toMatchSnapshot();
        });

        it('should render remove icon', () => {
            const { wrapper, instance } = setup(true, {
                index: set(1),
                empty: setFalse()
            });
            const removeIcon = getSpecWrapper(wrapper, removeIcoSpecId);

            expect(removeIcon.hasClass('removeIco')).toBeTruthy();
            expect(removeIcon.props()).toEqual(
                expect.objectContaining({
                    onClick: instance.onRemove
                })
            );
            expect(wrapper.find('.fe-remove').exists()).toBeTruthy();
            expect(wrapper).toMatchSnapshot();
        });
    });

    describe('instance methods', () => {
        describe('handleChangeName', () => {
            it('should update state, call onChange prop', () => {
                const setStateSpy = spyOn('setState');
                const { wrapper, props: { onChange: onChangeMock }, instance } = setup(true, {
                    onChange: setMock()
                });

                instance.handleChangeName(headers[0].name);
                wrapper.update();

                expect(setStateSpy).toHaveBeenCalledTimes(1);
                expect(setStateSpy).toHaveBeenCalledWith(
                    { name: headers[0].name },
                    expect.any(Function)
                );
                expect(onChangeMock).toHaveBeenCalledTimes(1);
                expect(onChangeMock).toHaveBeenCalledWith(expect.any(HeaderElement));

                expect(
                    wrapper
                        .find('Connect(TextInputElement)')
                        .at(0)
                        .prop('entry')
                ).toEqual({ value: headers[0].name });
                expect(wrapper).toMatchSnapshot();

                setStateSpy.mockRestore();
            });
        });

        describe('onChangeValue', () => {
            it('should update state, call onChange prop', () => {
                const setStateSpy = spyOn('setState');
                const { wrapper, props, instance } = setup(true, {
                    onChange: setMock()
                });

                instance.handleChangeValue(headers[0].value);
                wrapper.update();

                expect(setStateSpy).toHaveBeenCalledTimes(1);
                expect(setStateSpy).toHaveBeenCalledWith(
                    { value: headers[0].value },
                    expect.any(Function)
                );
                expect(props.onChange).toHaveBeenCalledTimes(1);
                expect(props.onChange).toHaveBeenCalledWith(expect.any(HeaderElement));
                expect(
                    wrapper
                        .find('Connect(TextInputElement)')
                        .at(1)
                        .prop('entry')
                ).toEqual({ value: headers[0].value });
                expect(wrapper).toMatchSnapshot();

                setStateSpy.mockRestore();
            });
        });
    });

    describe('onRemove', () => {
        it('should call onRemove prop', () => {
            const { wrapper, props, instance } = setup(true, {
                onRemove: setMock()
            });

            instance.onRemove();

            expect(props.onRemove).toHaveBeenCalledTimes(1);
        });
    });
});
