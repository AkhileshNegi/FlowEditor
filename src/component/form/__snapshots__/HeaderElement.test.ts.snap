// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderElement instance methods onChangeName should update state, call onChange prop 1`] = `
<FormElement
  errors={Array []}
  name="header_00c4498e-1c9e-4c26-aa7c-f81d13573129"
>
  <div
    className="header"
    data-spec="header-container"
  >
    <div
      className="header_name"
      data-spec="name-container"
    >
      <Connect(TextInputElement)
        name="name"
        onChange={[Function]}
        placeholder="Header Name"
        showInvalid={false}
        value="Content-Type"
      />
    </div>
    <div
      className="header_value"
      data-spec="value-container"
    >
      <Connect(TextInputElement)
        autocomplete={true}
        name="value"
        onChange={[Function]}
        placeholder="Value"
        value="application/json"
      />
    </div>
  </div>
</FormElement>
`;

exports[`HeaderElement instance methods onChangeValue should update state, call onChange prop 1`] = `
<FormElement
  errors={Array []}
  name="header_00c4498e-1c9e-4c26-aa7c-f81d13573129"
>
  <div
    className="header"
    data-spec="header-container"
  >
    <div
      className="header_name"
      data-spec="name-container"
    >
      <Connect(TextInputElement)
        name="name"
        onChange={[Function]}
        placeholder="Header Name"
        showInvalid={false}
        value="Content-Type"
      />
    </div>
    <div
      className="header_value"
      data-spec="value-container"
    >
      <Connect(TextInputElement)
        autocomplete={true}
        name="value"
        onChange={[Function]}
        placeholder="Value"
        value="application/json"
      />
    </div>
  </div>
</FormElement>
`;

exports[`HeaderElement render should render remove icon 1`] = `
<FormElement
  errors={Array []}
  name="header_00c4498e-1c9e-4c26-aa7c-f81d13573129"
>
  <div
    className="header"
    data-spec="header-container"
  >
    <div
      className="header_name"
      data-spec="name-container"
    >
      <Connect(TextInputElement)
        name="name"
        onChange={[Function]}
        placeholder="Header Name"
        showInvalid={false}
        value="Content-Type"
      />
    </div>
    <div
      className="header_value"
      data-spec="value-container"
    >
      <Connect(TextInputElement)
        autocomplete={true}
        name="value"
        onChange={[Function]}
        placeholder="Value"
        value="application/json"
      />
    </div>
    <div
      className="removeIco"
      data-spec="remove-icon"
      onClick={[Function]}
    >
      <span
        className="icn-remove"
      />
    </div>
  </div>
</FormElement>
`;

exports[`HeaderElement render should render self, children with base props 1`] = `
<FormElement
  errors={Array []}
  name="header_00c4498e-1c9e-4c26-aa7c-f81d13573129"
>
  <div
    className="header"
    data-spec="header-container"
  >
    <div
      className="header_name"
      data-spec="name-container"
    >
      <Connect(TextInputElement)
        name="name"
        onChange={[Function]}
        placeholder="Header Name"
        showInvalid={false}
        value="Content-Type"
      />
    </div>
    <div
      className="header_value"
      data-spec="value-container"
    >
      <Connect(TextInputElement)
        autocomplete={true}
        name="value"
        onChange={[Function]}
        placeholder="Value"
        value="application/json"
      />
    </div>
  </div>
</FormElement>
`;
